x/erc20/keeper/precompiles.go

```diff
@@ -65,3 +72,74 @@ func (k Keeper) RegisterERC20Extensions(ctx sdk.Context) error {
 	// add the ERC20s to the EVM active and available precompiles
 	return k.evmKeeper.AddEVMExtensions(ctx, precompiles...)
 }
+
+// RegisterTokenPairForNativeCoin creates a token pair for the native coin for a newly seen denom together with a
+// precompile instance for the token pair.
+func (k Keeper) RegisterTokenPairForNativeCoin(
+	ctx sdk.Context,
+	coinMetadata banktypes.Metadata,
+) (*types.TokenPair, error) {
+	// Check if denomination is already registered
+	if k.IsDenomRegistered(ctx, coinMetadata.Name) {
+		return nil, errorsmod.Wrapf(
+			types.ErrTokenPairAlreadyExists, "coin denomination already registered: %s", coinMetadata.Name,
+		)
+	}
+
+	// Check if the coin exists by ensuring the supply is set
+	if !k.bankKeeper.HasSupply(ctx, coinMetadata.Base) {
+		return nil, errorsmod.Wrapf(
+			errortypes.ErrInvalidCoins, "base denomination '%s' cannot have a supply of 0", coinMetadata.Base,
+		)
+	}
+
+	if err := k.verifyMetadata(ctx, coinMetadata); err != nil {
+		return nil, errorsmod.Wrapf(
+			types.ErrInternalTokenPair, "coin metadata is invalid %s", coinMetadata.Name,
+		)
+	}
+
+	prefix := transfertypes.DenomPrefix + "/"
+	if len(coinMetadata.Base) < len(prefix) {
+		return nil, errorsmod.Wrapf(transfertypes.ErrInvalidDenomForTransfer, "invalid denom %s", coinMetadata.Base)
+	}
+
+	hexBz, err := hex.DecodeString(coinMetadata.Base[len(prefix):])
+	if err != nil {
+		return nil, errorsmod.Wrapf(transfertypes.ErrInvalidDenomForTransfer, "invalid hex %s", coinMetadata.Base)
+	}
+
+	addr := common.BytesToAddress(hexBz)
+
+	// TODO: Here we need to move the truncation to 20 ?
+	pair := types.NewTokenPair(addr, coinMetadata.Base, types.OWNER_MODULE)
+
+	k.SetTokenPair(ctx, pair)
+	k.SetDenomMap(ctx, pair.Denom, pair.GetID())
+	k.SetERC20Map(ctx, common.HexToAddress(pair.Erc20Address), pair.GetID())
+
+	return &pair, nil
+}
+
+// RegisterPrecompileForCoin Adds an erc20 precompile interface for an IBC Coin.
+// It truncates the denom address to 20 bytes and registers the precompile if it is not already registered
+func (k Keeper) RegisterPrecompileForCoin(ctx sdk.Context, pair types.TokenPair) error {
+	params := k.evmKeeper.GetParams(ctx)
+	found := params.IsPrecompileRegistered(pair.Erc20Address[:20])
+	if found {
+		return nil
+	}
+
+	// Register a new precompile address
+	newPrecompile, err := erc20.NewPrecompile(pair, k.bankKeeper, k.authzKeeper, *k.transferKeeper)
+	if err != nil {
+		return err
+	}
+
+	err = k.evmKeeper.AddEVMExtensions(ctx, newPrecompile)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
```

Comment:
This is a separate function to create the precompile and add it to EVM extensions